dbms/src/Interpreters/ExpressionJIT.cpp

```diff
@@ -0,0 +1,318 @@
+        if (typeIsA<DataTypeInt64>(type) || typeIsA<DataTypeUInt64>(type))
+            return builder.getInt64Ty();
+        if (typeIsA<DataTypeFloat32>(type))
+            return builder.getFloatTy();
+        if (typeIsA<DataTypeFloat64>(type))
+            return builder.getDoubleTy();
+        return nullptr;
+    }
+
+    LLVMCompiledFunction * lookup(const std::string& name)
+    {
+        std::string mangledName;
+        llvm::raw_string_ostream mangledNameStream(mangledName);
+        llvm::Mangler::getNameWithPrefix(mangledNameStream, name, layout);
+        /// why is `findSymbol` not const? we may never know.
+        return reinterpret_cast<LLVMCompiledFunction *>(compileLayer.findSymbol(mangledNameStream.str(), false).getAddress().get());
+    }
+};
+
+LLVMContext::LLVMContext()
+    : shared(std::make_shared<LLVMContext::Data>())
+{}
+
+void LLVMContext::finalize()
+{
+    if (!shared->module->size())
+        return;
+    shared->module->print(llvm::errs(), nullptr, false, true);
+    llvm::PassManagerBuilder builder;
+    llvm::legacy::FunctionPassManager fpm(shared->module.get());
+    builder.OptLevel = 2;
+    builder.populateFunctionPassManager(fpm);
+    for (auto & function : *shared->module)
+        fpm.run(function);
+    llvm::cantFail(shared->compileLayer.addModule(shared->module, std::make_shared<llvm::orc::NullResolver>()));
+    shared->module->print(llvm::errs(), nullptr, false, true);
+}
+
+bool LLVMContext::isCompilable(const IFunctionBase& function) const
+{
+    if (!function.isCompilable() || !shared->toNativeType(function.getReturnType()))
+        return false;
+    for (const auto & type : function.getArgumentTypes())
+        if (!shared->toNativeType(type))
+            return false;
+    return true;
+}
+
+LLVMPreparedFunction::LLVMPreparedFunction(LLVMContext context, std::shared_ptr<const IFunctionBase> parent)
+    : parent(parent), context(context), function(context->lookup(parent->getName()))
+{}
+
+static MutableColumnPtr createNonNullableColumn(const DataTypePtr & type)
+{
+    if (auto * nullable = typeid_cast<const DataTypeNullable *>(type.get()))
+        return createNonNullableColumn(nullable->getNestedType());
+    return type->createColumn();
+}
+
+void LLVMPreparedFunction::executeImpl(Block & block, const ColumnNumbers & arguments, size_t result)
+{
+    size_t block_size = 0;
+    std::vector<const void *> columns(arguments.size());
+    std::vector<char> is_const(arguments.size());
+    for (size_t i = 0; i < arguments.size(); i++)
+    {
+        auto * column = block.getByPosition(arguments[i]).column.get();
+        if (column->size())
+            /// assume the column is a `ColumnVector<T>`. there's probably no good way to actually
+            /// check that at runtime, so let's just hope it's always true for columns containing types
+            /// for which `LLVMContext::Data::toNativeType` returns non-null.
+            columns[i] = column->getDataAt(0).data;
```

Comment:
`IColumn::isFixedAndContiguous` seems to be what I wanted here. The compiled loop could be extended to arbitrary columns for which this method returns true passing a tuple of strides (e.g. string length for `ColumnFixedString`) instead of a tuple of "is constant" flags, though I'm not sure how this could affect loop auto-vectorization.